### PR TITLE
Feature/event bus

### PR DESCRIPTION
**Pull request by zhenlei520**
1. Local message table retry
2. UoW adjustment

### CODE REVIEW

File: src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.Dapr/Internal/LocalQueueProcessor.cs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        catch (Exception ex)
+        {
+            Logger.LogWarning(ex, "... getting local retry queue error");
+
+            Task.Delay(TimeSpan.FromSeconds(2));
+            return new List<IntegrationEventLogItems>();
+        }
+    }
+
+    private void SafeMethods(Action action, string name)
```
Comment (doddgu) [inline src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.Dapr/Internal/LocalQueueProcessor.cs:63]: Does just ignore exception called safe?

File: src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.Dapr/Options/DispatcherOptions.cs [hunk 2]
```diff
@@ -13,13 +13,63 @@ public string PubSubName
...
+
+    /// <summary>
+    /// The interval at which the local queue is polled for failed messages.
+    /// Local queue does not rebuild after service crash
+    /// Default is 3 seconds.
+    /// </summary>
+    public int LocalFailedRetryInterval { get; set; } = 3;
+
+    /// <summary>
+    /// The size of a single event to be retried
```
Comment (doddgu) [inline src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.Dapr/Options/DispatcherOptions.cs:47]: What's the mean?

File: src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.Dapr/Options/DispatcherOptions.cs [hunk 3]
```diff
@@ -13,13 +13,63 @@ public string PubSubName
...
+    /// Delete expired event interval
+    /// Default is 300 seconds.
+    /// </summary>
+    public int CleaningExpireInterval { get; set; } = 300;
+
+    /// <summary>
+    /// Expiration time, when the message status is successful and has expired, it will be deleted by the scheduled task
+    /// Default: ( 24 * 3600 )s
+    /// </summary>
+    public long ExpireDate { get; set; } = 24 * 3600;
```
Comment (doddgu) [inline src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.Dapr/Options/DispatcherOptions.cs:68]: ExpireDate -> PublishedExpireTime

File: src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.Dapr/Processor/DeleteDataExpiresProcessor.cs [hunk 4]
```diff
@@ -0,0 +1,37 @@
+﻿namespace MASA.Contrib.Dispatcher.IntegrationEvents.Dapr.Processor;
+
+public class DeleteDataExpiresProcessor : IProcessor
```
Comment (doddgu) [inline src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.Dapr/Processor/DeleteDataExpiresProcessor.cs:3]: DeleteDataExpiresProcessor -> DeletePublishedExpireEventProcessor

File: src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.Dapr/Processor/DeleteDataExpiresProcessor.cs [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    /// </summary>
+    /// <param name="stoppingToken"></param>
+    /// <returns></returns>
+    public async Task ExecuteAsync(CancellationToken stoppingToken)
+    {
+        using (var scope = _serviceProvider.CreateScope())
+        {
+            var logService = scope.ServiceProvider.GetRequiredService<IIntegrationEventLogService>();
+            var expireDate = DateTime.Now.AddSeconds(-_options.Value.ExpireDate);
+
```
Comment (doddgu) [inline src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.Dapr/Processor/DeleteDataExpiresProcessor.cs:30]: Too more blank line.

File: src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.Dapr/Processor/DeleteDataExpiresProcessor.cs [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+            //todo: Delete expired events
+        }
+        await Task.Delay(_options.Value.CleaningExpireInterval, stoppingToken);
+        //todo: Delete successfully published and expired messages
```
Comment (doddgu) [inline src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.Dapr/Processor/DeleteDataExpiresProcessor.cs:35]: If you don't finish local message, you can split them to another pr.

File: src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.Dapr/Processor/DeleteLocalQueueExpiresProcessor.cs [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+    /// <summary>
+    /// Delete expired events
+    /// </summary>
+    /// <param name="stoppingToken"></param>
+    /// <returns></returns>
+    public async Task ExecuteAsync(CancellationToken stoppingToken)
+    {
+        LocalQueueProcessor.Default.DeleteAsync(_options.Value.LocalRetryTimes);
+        await Task.Delay(_options.Value.CleaningLocalQueueExpireInterval, stoppingToken);
```
Comment (doddgu) [inline src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.Dapr/Processor/DeleteLocalQueueExpiresProcessor.cs:20]: Why call delay here? This is a very bad design!

Comment (doddgu) [reply src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.Dapr/Processor/DeleteLocalQueueExpiresProcessor.cs:20]: If you want use `InfiniteLoopProcessor`, you can add `Sleep` method in IProcessor, then in `InfiniteLoopProcessor` call `_processor.Sleep()`

File: src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.Dapr/Internal/LocalQueueProcessor.cs [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    public List<IntegrationEventLogItems> RetrieveEventLogsFailedToPublishAsync(int maxRetryTimes)
+    {
+        try
+        {
+            return _retryEventLogs
+                .Select(item => item.Value)
+                .Where(log => log.RetryCount < maxRetryTimes)
+                .OrderBy(log => log.RetryCount)
+                .ThenBy(log => log.CreationTime)
+                .ToList();
```
[the remark this answers is not in the document]
Comment (doddgu) [reply src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.Dapr/Internal/LocalQueueProcessor.cs:52]: RetryBatchSize?

File: src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.Dapr/README.zh-CN.md [hunk 9]
```diff
@@ -65,3 +65,41 @@ public async Task DomeIntegrationEventHandleAsync(DomeIntegrationEvent @event)
...
+
+持久化数据源队列：
+
+特点：
+
+- 系统崩溃后，可以从db或者其他持久化源获取重试队列，确保事件100%重试
+- 作为本地内存队列的降级方案，对db或者其他数据源压力更低
+
+在单副本情况下，两种队列的任务仅会在单个队列中执行，不会存在两个队列同时执行的情况。
+在多副本情况下，同一个任务可能会被多个副本所执行，虽然我们有做幂等，但为交付保证是 At Least Once，仍然有可能出现事件发布成功，但状态更改失败的情况，
```
Comment (doddgu) [inline src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.Dapr/README.zh-CN.md:103]: 在多副本情况下，同一个任务可能会被多个副本所执行。

这个意思是所有副本都会重试？没有分配压力？

File: src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.Dapr/Servers/DefaultHostedService.cs [hunk 10]
```diff
@@ -0,0 +1,20 @@
+﻿namespace MASA.Contrib.Dispatcher.IntegrationEvents.Dapr.Servers;
+
+public class DefaultHostedService : IProcessingServer
```
Comment (doddgu) [inline src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.Dapr/Servers/DefaultHostedService.cs:3]: IProcessingServer? or BackgroundService enough?

File: src/DDD/MASA.Contrib.DDD.Domain.Repository.EF/Repository.cs [hunk 11]
```diff
@@ -110,52 +129,78 @@ public override Task<List<TEntity>> GetPaginatedListAsync(int skip, int take, Di
...
         _context.Set<TEntity>().Update(entity);
+        UnitOfWork.EntityState = EntityState.Changed;
+        CheckAndOpenTransaction();
         return Task.FromResult(entity);
     }
 
     public override Task UpdateRangeAsync(IEnumerable<TEntity> entities, CancellationToken cancellationToken = default)
     {
         _context.Set<TEntity>().UpdateRange(entities);
+        UnitOfWork.EntityState = EntityState.Changed;
```
[the remark this answers is not in the document]
Comment (zhenlei520) [reply src/DDD/MASA.Contrib.DDD.Domain.Repository.EF/Repository.cs:188]: You are right, the BaseRepository implementation of IUnitOfWork is missing

File: src/DDD/MASA.Contrib.DDD.Domain/Events/DomainCommand.cs [hunk 12]
```diff
@@ -1,21 +1,9 @@
 namespace MASA.Contrib.DDD.Domain.Events;
 
-public record DomainCommand : IDomainCommand
+public record DomainCommand(Guid Id, DateTime CreationTime) : IDomainCommand
```
[the remark this answers is not in the document]
Comment (zhenlei520) [reply src/DDD/MASA.Contrib.DDD.Domain/Events/DomainCommand.cs:3]: This block has two constructors by default, a constructor with no parameters, and a constructor with two parameters. UnitOfWork is not required. It is assigned by the framework, not through the constructor. ?

File: src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.Dapr/Processor/DeleteDataExpiresProcessor.cs (hunk 6 above)
Comment (zhenlei520) [reply src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.Dapr/Processor/DeleteDataExpiresProcessor.cs:35]: Deleting expired messages has been implemented

File: src/BasicAbility/MASA.Contrib.BasicAbility.Dcc/ConfigurationAPIClient.cs [hunk 13]
```diff
@@ -105,7 +105,7 @@ private async Task<(string Raw, ConfigurationTypes ConfigurationType)> GetRawByK
         return FormatRaw(raw);
     }
 
-    private (string Raw, ConfigurationTypes ConfigurationType) FormatRaw(string raw)
+    private (string Raw, ConfigurationTypes ConfigurationType) FormatRaw(string? raw)
```
[the remark this answers is not in the document]
Comment (zhenlei520) [reply src/BasicAbility/MASA.Contrib.BasicAbility.Dcc/ConfigurationAPIClient.cs:108]: It is recorded here as ConfigurationAPIClient instead of ConfigurationApiClient because the full name of API is Application Programming Interface, so it is not named after a camel

File: src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.Dapr/Processor/DeleteLocalQueueExpiresProcessor.cs (hunk 7 above)
Comment (zhenlei520) [reply src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.Dapr/Processor/DeleteLocalQueueExpiresProcessor.cs:20]: good idea, already adjusted

File: src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.Dapr/README.zh-CN.md (hunk 9 above)
Comment (zhenlei520) [reply src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.Dapr/README.zh-CN.md:103]: This is controlled by the concurrent lock and State of RowVersion. If the task is processed by other instances, it will be skipped, but it is not sure whether there will be situations that are not considered to cause the task to be repeated.

File: src/DDD/MASA.Contrib.DDD.Domain/Events/DomainCommand.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (doddgu) [reply src/DDD/MASA.Contrib.DDD.Domain/Events/DomainCommand.cs:3]: `JsonIgnore`

File: src/Data/MASA.Contrib.Data.UoW.EF/UnitOfWork.cs [hunk 14]
```diff
@@ -21,6 +21,8 @@ public DbTransaction Transaction
 
     public bool DisableRollbackOnFailure { get; set; }
 
+    public EntityState EntityState { get; set; }
+
```
[the remark this answers is not in the document]
Comment (zhenlei520) [reply src/Data/MASA.Contrib.Data.UoW.EF/UnitOfWork.cs:25]: processed

File: src/Dispatcher/MASA.Contrib.Dispatcher.Events/Internal/Middleware/TransactionMiddleware.cs [hunk 15]
```diff
@@ -3,30 +3,44 @@ namespace MASA.Contrib.Dispatcher.Events.Internal.Middleware;
...
                 await transaction!.UnitOfWork!.CommitAsync();
             }
         }
         catch (Exception)
         {
-            if (IsUseTransaction(@event, out ITransaction? transaction))
+            if (IsUseTransaction(@event, out ITransaction? transaction) && !transaction!.UnitOfWork!.DisableRollbackOnFailure)
             {
-                await transaction!.UnitOfWork!.RollbackAsync();
+                await transaction.UnitOfWork!.RollbackAsync();
```
[the remark this answers is not in the document]
Comment (zhenlei520) [reply src/Dispatcher/MASA.Contrib.Dispatcher.Events/Internal/Middleware/TransactionMiddleware.cs:32]: transaction must not be null, ! is redundant, so removed

File: src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.Dapr/Internal/LocalQueueProcessor.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (zhenlei520) [reply src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.Dapr/Internal/LocalQueueProcessor.cs:52]: The local queue query cycle is very short, so this block does not do RetryBatchSize, I will modify this block to support RetryBatchSize

File: src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.Dapr/Options/DispatcherOptions.cs (hunk 2 above)
Comment (zhenlei520) [reply src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.Dapr/Options/DispatcherOptions.cs:47]: The number of retry messages obtained for each retry

File: src/DDD/MASA.Contrib.DDD.Domain.Repository.EF/Repository.cs [hunk 16]
```diff
@@ -110,52 +129,78 @@ public override Task<List<TEntity>> GetPaginatedListAsync(int skip, int take, Di
...
 
     public override Task RollbackAsync(CancellationToken cancellationToken = default)
         => UnitOfWork.RollbackAsync(cancellationToken);
 
-    public override Task SaveChangesAsync(CancellationToken cancellationToken = default)
-         => UnitOfWork.SaveChangesAsync(cancellationToken);
+    public override async Task SaveChangesAsync(CancellationToken cancellationToken = default)
+    {
+        await UnitOfWork.SaveChangesAsync(cancellationToken);
+        UnitOfWork.EntityState = EntityState.Unchanged;
```
[the remark this answers is not in the document]
Comment (zhenlei520) [reply src/DDD/MASA.Contrib.DDD.Domain.Repository.EF/Repository.cs]: processed

File: src/Dispatcher/MASA.Contrib.Dispatcher.Events/Internal/Middleware/TransactionMiddleware.cs [hunk 17]
```diff
@@ -3,30 +3,44 @@ namespace MASA.Contrib.Dispatcher.Events.Internal.Middleware;
...
+        _unitOfWork = unitOfWork;
+    }
+
     public async Task HandleAsync(TEvent @event, EventHandlerDelegate next)
     {
         try
         {
             await next();
 
+            if (_unitOfWork != null && _unitOfWork.EntityState == EntityState.Changed)
```
[the remark this answers is not in the document]
Comment (zhenlei520) [reply src/Dispatcher/MASA.Contrib.Dispatcher.Events/Internal/Middleware/TransactionMiddleware.cs]: _unitOfWork is { EntityState: EntityState.Changed }

File: src/Dispatcher/MASA.Contrib.Dispatcher.Events/README.zh-CN.md [hunk 18]
```diff
@@ -172,4 +172,23 @@ builder.Services
...
-> 配合Contracts.EF、UnitOfWork使用，当Event实现了ITransaction，会在执行第一次CUD后自动开启事务，且在Handler全部执行后提交事务，当事务出现异常后，会自动回滚事务
+> 配合MASA.Contrib.DDD.Domain.Repository.EF.Repository、UnitOfWork使用，当Event实现了ITransaction，会在执行Add、Update、Delete方法时自动开启事务，且在Handler全部执行后提交事务，当事务出现异常后，会自动回滚事务
+
+##### 总结
+
+IEventBus是事件总线的核心，配合Cqrs、Uow、MASA.Contrib.DDD.Domain.Repository.EF使用，可实现自动执行SaveChange（启用UoW）与Commit（启用UoW且无关闭事务）操作，并支持出现异常后，回滚事务
+
+> 问题1. 通过eventBus发布事件，Handler出错，但数据已经保存到数据库中，事务并未回滚
+
+    > 1. 检查自定义事件是否实现ITransaction或继承类是否有实现ITransaction
```
[the remark this answers is not in the document]
Comment (zhenlei520) [reply src/Dispatcher/MASA.Contrib.Dispatcher.Events/README.zh-CN.md]: processed

File: src/Dispatcher/MASA.Contrib.Dispatcher.Events/README.zh-CN.md [hunk 19]
```diff
@@ -172,4 +172,23 @@ builder.Services
...
+    > 2. 是否使用UoW
```
[the remark this answers is not in the document]
Comment (zhenlei520) [reply src/Dispatcher/MASA.Contrib.Dispatcher.Events/README.zh-CN.md]: processed

File: src/Dispatcher/MASA.Contrib.Dispatcher.Events/README.zh-CN.md [hunk 20]
```diff
@@ -172,4 +172,23 @@ builder.Services
...
+    > 3. 检查UnitOfWork的UseTransaction属性是否为false
```
[the remark this answers is not in the document]
Comment (zhenlei520) [reply src/Dispatcher/MASA.Contrib.Dispatcher.Events/README.zh-CN.md]: processed

File: src/Dispatcher/MASA.Contrib.Dispatcher.Events/README.zh-CN.md [hunk 21]
```diff
@@ -172,4 +172,23 @@ builder.Services
...
+    > 4. 检查UnitOfWork的DisableRollbackOnFailure属性是否为true
```
[the remark this answers is not in the document]
Comment (zhenlei520) [reply src/Dispatcher/MASA.Contrib.Dispatcher.Events/README.zh-CN.md]: processed

File: src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.Dapr/Options/DispatcherOptions.cs (hunk 2 above)
Comment (doddgu) [reply src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.Dapr/Options/DispatcherOptions.cs]: Maximum number of retries?

File: src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.Dapr/Processor/InfiniteLoopProcessor.cs [hunk 22]
```diff
@@ -0,0 +1,37 @@
+﻿namespace MASA.Contrib.Dispatcher.IntegrationEvents.Dapr.Processor;
+
+public class InfiniteLoopProcessor : ProcessorBase, IProcessor
```
Comment (doddgu) [inline src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.Dapr/Processor/InfiniteLoopProcessor.cs:3]: Remove the IProcessor

File: src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.Dapr/Processor/RetryByDataProcessor.cs [hunk 23]
```diff
@@ -0,0 +1,76 @@
+﻿namespace MASA.Contrib.Dispatcher.IntegrationEvents.Dapr.Processor;
+
+public class RetryByDataProcessor : ProcessorBase, IProcessor
```
Comment (doddgu) [inline src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.Dapr/Processor/RetryByDataProcessor.cs:3]: Remove the IProcessor

File: src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.Dapr/README.md [hunk 24]
```diff
@@ -64,4 +64,41 @@ public async Task DomeIntegrationEventHandleAsync(DomeIntegrationEvent @event)
 {
     //todo
 }
-```
\ No newline at end of file
+```
+
+### retry strategy
```
Comment (doddgu) [inline src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.Dapr/README.md:69]: strategy -> policy

File: src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.EventLogs.EF/IntegrationEventLogService.cs [hunk 25]
```diff
@@ -48,32 +62,84 @@ public async Task SaveEventAsync(IIntegrationEvent @event, DbTransaction transac
...
     {
-        return UpdateEventStatus(eventId, IntegrationEventStates.Published);
+        return UpdateEventStatus(eventId, IntegrationEventStates.Published, eventLog =>
+        {
+            if (eventLog.State != IntegrationEventStates.InProgress)
+            {
+                _logger?.LogWarning(
+                    "Failed to modify the state of the local message table to {OptState}, the current State is {State}, Id: {Id}",
+                    IntegrationEventStates.Published, eventLog.State, eventLog.Id);
+                throw new UserFriendlyException("Failed to change state");
```
Comment (doddgu) [inline src/Dispatcher/MASA.Contrib.Dispatcher.IntegrationEvents.EventLogs.EF/IntegrationEventLogService.cs:71]: Make the exception message more detailed.

**Review by doddgu — APPROVED** [review]
Submit a version for now
